tempesta_fw/connection.c

```diff
@@ -50,32 +54,39 @@ tfw_connection_init(TfwConnection *conn)
 
 	INIT_LIST_HEAD(&conn->list);
 	INIT_LIST_HEAD(&conn->msg_queue);
+	spin_lock_init(&conn->msg_qlock);
+	spin_lock_init(&conn->splock);
+	atomic_set(&conn->refcnt, 1);
 }
 
 /*
- * It's essential that this function is called before a socket
- * is bound or connected, which guarantees that there are no calls
- * to Tempesta callbacks. No locking is needed under these conditions.
- * Otherwise, it must be called under write lock on sk->sk_callback_lock.
+ * It's essential that this function is called before a socket is bound
+ * or connected, which guarantees that there are no calls to Tempesta
+ * callbacks. No locking is required under these conditions to modify
+ * sk->sk_user_data. Otherwise, it must be called under write lock on
+ * sk->sk_callback_lock. conn->sk is modified under a separate lock.
  */
 void
 tfw_connection_link_sk(TfwConnection *conn, struct sock *sk)
 {
-	BUG_ON(conn->sk || sk->sk_user_data);
-	conn->sk = sk;
-	sk->sk_user_data = conn;
+	tfw_connection_link_to_sk(conn, sk);
+	tfw_connection_link_from_sk(conn, sk);
 }
 
 /*
- * This function does the opposite to what tfw_connection_link_sk() does.
- * It must be called under write lock on sk->sk_callback_lock to be able
- * to modify sk->sk_user_data, or the socket must not be bound or connected.
+ * This function does an opposite to what tfw_connection_link_sk() does.
+ * It must be called under write lock on sk->sk_callback_lock to modify
+ * sk->sk_user_data, or the socket must not be bound or connected.
+ * Socket close may occur in SoftIRQ or in user context at STOP time,
+ * and in both cases it's protected with a lock on sk->sk_callback_lock.
+ * conn->sk is modified under a separate lock.
  */
 void
-tfw_connection_unlink_sk(TfwConnection *conn)
+tfw_connection_unlink_sk(TfwConnection *conn, struct sock *sk)
```

Comment:
The sk argument is used for BUG_ON statement only, so I think compiler complains to unused argument if you compile Tempesta w/o DEBUG option. Probably it's better to check sk in the caller function.